gallery_dl/extractor/fantia.py

```diff
@@ -0,0 +1,149 @@
+# -*- coding: utf-8 -*-
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 as
+# published by the Free Software Foundation.
+
+"""Extractors for https://fantia.jp/"""
+
+from .common import Extractor, Message
+from .. import text
+
+
+class FantiaExtractor(Extractor):
+    """Base class for Fantia extractors"""
+    category = "fantia"
+    root = "https://fantia.jp"
+    directory_fmt = ("{category}", "{fanclub_id}")
+    filename_fmt = "{post_id}_{file_id}.{extension}"
+    archive_fmt = "{post_id}_{file_id}"
+    _warning = True
+
+    def items(self):
+        yield Message.Version, 1
+
+        if self._warning:
+            if "_session_id" not in self.session.cookies:
+                self.log.warning("no '_session_id' cookie set")
+            FantiaExtractor._warning = False
+
+        for full_response, post in self.posts():
+            yield Message.Directory, post
+            for url, url_data in self._get_urls_from_post(full_response, post):
+                fname = url_data["content_filename"] or url
+                text.nameext_from_url(fname, url_data)
+                url_data["file_url"] = url
+                yield Message.Url, url, url_data
+
+    def posts(self):
+        """Return all relevant post objects"""
+
+    def _pagination(self, base_url):
+        headers = {"Referer": self.root}
+        page = 1
+        posts_found = True
+
+        while posts_found:
+            url = base_url+str(page)
+            url = text.ensure_http_scheme(url)
+            gallery_page_html = self.request(url, headers=headers).text
+            posts_found = False
+            for post_id in text.extract_iter(
+                gallery_page_html, 'class="link-block" href="/posts/', '"'
+            ):
+                posts_found = True
+                yield self._get_post_data(post_id)
+
+            page += 1
```

Comment:
I'd suggest using separate `url` and `params` instead of using a `base_url` and manually adding the current page. This could theoretically also be done for Fanbox, but we have `body["nextUrl"]` there.
I'd also only return post IDs from here (and all `posts()` methods including the one from `FantiaPostExtractor`) and call  `_get_post_data(post_id)` in the main loop. We can also get rid of `posts_found` and the call to `ensure_http_scheme()`.

```suggestion
def _pagination(self, url):
        params = {"page": 1}
        headers = {"Referer": self.root}

        while True:
            page = self.request(url, params=params, headers=headers).text

            post_id = None
            for post_id in text.extract_iter(
                    page, 'class="link-block" href="/posts/', '"'):
                yield post_id

            if not post_id:
                return
            params["page"] += 1
```